Mage.Sets/src/mage/cards/a/AnrakyrTheTraveller.java

```diff
@@ -0,0 +1,198 @@
+
+/**
+ * @author gravitybone
+ * @author Aquid
+ */
+public final class AnrakyrTheTraveller extends CardImpl {
+
+    private static final FilterCard filter = new FilterArtifactCard();
+
+
+    public AnrakyrTheTraveller(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[] { CardType.ARTIFACT, CardType.CREATURE }, "{4}{B}");
+
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.NECRON);
+        this.power = new MageInt(4);
+        this.toughness = new MageInt(4);
+
+        // Lord of the Pyrrhian Legions — Whenever Anrakyr the Traveller attacks, you may cast an artifact spell from your hand or graveyard by paying life equal to its mana value rather than paying its mana cost.
+        Ability ability = new AttacksTriggeredAbility(new AnrakyrTheTravellerEffect(filter));
+
+        this.addAbility(ability.withFlavorWord("Lord of the Pyrrhian Legions"));
+
+    }
+
+    private AnrakyrTheTraveller(final AnrakyrTheTraveller card) {
+        super(card);
+    }
+
+    @Override
+    public Card copy() {
+        return new AnrakyrTheTraveller(this);
+    }
+}
+
+class AnrakyrTheTravellerEffect extends OneShotEffect {
+    private final FilterCard filter;
+
+    public AnrakyrTheTravellerEffect(FilterCard filter) {
+        super(Outcome.AIDontUseIt);
+        this.filter = filter;
+        this.staticText = "you may cast an " + filter.getMessage() + " from your hand or graveyard by paying life equal to its mana value rather than paying its mana cost.";
+    }
+
+    protected AnrakyrTheTravellerEffect(final AnrakyrTheTravellerEffect effect) {
+        super(effect);
+        this.filter = effect.filter;
+    }
+
+    
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        if (player == null) {
+            return false;
+        }
+        
+        Set<Card> cards = player.getHand().getCards(filter, game);
+        cards.addAll(player.getGraveyard().getCards(filter, game));
+        
+        Map<UUID, List<Card>> cardMap = new HashMap<>();
+        for (Card card : cards) {
+            List<Card> castableComponents = getCastableComponents(card, filter, source, player, game, false);
+            if (!castableComponents.isEmpty()) {
+                cardMap.put(card.getId(), castableComponents);
+            }
+        }      
+        Card cardToCast;
+        switch (cardMap.size()) {
+            case 0:
+                return false;
+            default:
+                Cards castableCards = new CardsImpl(cardMap.keySet());
+                TargetCard target = new TargetCard(0, 1, Zone.ALL, filter);
+                target.withNotTarget(true);
+                player.choose(Outcome.Benefit, castableCards, target, source, game);
+                cardToCast = castableCards.get(target.getFirstTarget(), game);
+        }  
+
+        if (cardToCast == null) {
```

Comment:
So that is my personal opinion, but I think you would want a prompt to continue if you have nothing available to cast with the ability. Otherwise the ability resolving automatically makes it obvious that you have no artifact in hand.
I'm not sure we are very clean on those kind of small information leak in general.
(Would also apply to `cardMap` being empty.)

Reply:
How do you usually invoke such a prompt? We tried `player.choice()` but that did not seem quiet right.

Reply:
I think that `player.chooseUse` should get you there.

Reply:
There should be a choice already due to the ability being optional- should not need to add another in the custom effect

Reply:
We changed the ability to be optional in [bab6ba3](https://github.com/magefree/mage/pull/11160/commits/bab6ba3c483930b806df9789d7caeeb626e63b90), please verify if it resolves this issue.